scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatWriter.scala

```diff
@@ -258,13 +258,31 @@ class FormatWriter(formatOps: FormatOps) {
       case x => x
     }
     initStyle.newlines.alwaysBeforeTopLevelStatements && {
-      topLevelTokens.contains(hash(toks(i).formatToken.right)) && {
-        val (distance, FormatToken(_, nextNonComment, _)) =
-          nextNonCommentWithCount(toks(i).formatToken)
-        isMultiline(actualOwner(nextNonComment).tokens.last, i + distance + 1)
+      val formatToken = toks(i).formatToken
+      def checkPackage: Option[Boolean] = {
```

Comment:
We should guard this change against `edition = 2019-11`

Reply:
will add the edition here, and post the diffs from scala-repos later today.

for now, working on #1557, i think the solution is incomplete, but chains are surprisingly difficult.

Reply:
Chains are very difficult with the current router design. I spent a LOT of time trying to break on plain `a.b` without exploding the search state but failed.